### PR TITLE
Agent: support unstable-codegen access via SOCKS proxy

### PR DESCRIPTION
**Pull request by cbart**
Part of https://github.com/sourcegraph/sourcegraph/issues/56254

Add `cody.proxy` that makes agent connect with a SOCKS proxy for `unstable-codegen`.

This change paired with https://github.com/sourcegraph/sourcegraph/pull/56264 routes autocomplete LLM traffic through a SOCKS proxy by extending `~/.sourcegraph-jetbrains.properties` :

```
cody.proxy: socks5://127.0.0.1:9999
```

For VSCode the setting is not shown in the UI, not suggested as an autocomplete option in _settings.json_, and when typed manually, the deprecation message is shown (https://github.com/sourcegraph/cody/pull/836#discussion_r1318212176).

Screenshot 2023-09-07 at 10 34 48

## Test plan

- [x] Manually tested with IntelliJ + agent
- [x] Manually tested with VS Code (using http.proxy as described by the config description)

### CODE REVIEW

**Comment by cbart** [comment]
I think I might be a little confused with dependency management here. Do I need to (and if so how do I) add transitive (?) dependencies of newly added socks-proxy?

```
➜  cody git:(cb/s/56254) pnpm -C vscode run build

> cody-ai@0.8.0 build /Users/cbart/cody/vscode
> tsc --build && pnpm run -s _build:esbuild:desktop && pnpm run -s _build:esbuild:web && pnpm run -s _build:webviews --mode production


  dist/extension.node.js      6.1mb ⚠️
  dist/extension.node.js.map  9.8mb

⚡ Done in 160ms
✘ [ERROR] Could not resolve "dns"

    ../node_modules/.pnpm/socks-proxy-agent@8.0.1/node_modules/socks-proxy-agent/dist/index.js:33:33:
      33 │ const dns = __importStar(require("dns"));
         ╵                                  ~~~~~

  The package "dns" wasn't found on the file system but is built into node. Are you trying to bundle
  for node? You can use "--platform=node" to do that, which will remove this error.

✘ [ERROR] Could not resolve "net"

    ../node_modules/.pnpm/socks-proxy-agent@8.0.1/node_modules/socks-proxy-agent/dist/index.js:34:33:
      34 │ const net = __importStar(require("net"));
         ╵                                  ~~~~~

  The package "net" wasn't found on the file system but is built into node. Are you trying to bundle
  for node? You can use "--platform=node" to do that, which will remove this error.

✘ [ERROR] Could not resolve "tls"

    ../node_modules/.pnpm/socks-proxy-agent@8.0.1/node_modules/socks-proxy-agent/dist/index.js:35:33:
      35 │ const tls = __importStar(require("tls"));
         ╵                                  ~~~~~

  The package "tls" wasn't found on the file system but is built into node. Are you trying to bundle
  for node? You can use "--platform=node" to do that, which will remove this error.

✘ [ERROR] Could not resolve "http"

    ../node_modules/.pnpm/agent-base@7.1.0/node_modules/agent-base/dist/index.js:30:34:
      30 │ const http = __importStar(require("http"));
         ╵                                   ~~~~~~

  The package "http" wasn't found on the file system but is built into node. Are you trying to
  bundle for node? You can use "--platform=node" to do that, which will remove this error.

✘ [ERROR] Could not resolve "http"

    ../node_modules/.pnpm/agent-base@7.1.0/node_modules/agent-base/dist/helpers.js:27:34:
      27 │ const http = __importStar(require("http"));
         ╵                                   ~~~~~~

  The package "http" wasn't found on the file system but is built into node. Are you trying to
  bundle for node? You can use "--platform=node" to do that, which will remove this error.

✘ [ERROR] Could not resolve "https"

    ../node_modules/.pnpm/agent-base@7.1.0/node_modules/agent-base/dist/helpers.js:28:35:
      28 │ const https = __importStar(require("https"));
         ╵                                    ~~~~~~~

  The package "https" wasn't found on the file system but is built into node. Are you trying to
  bundle for node? You can use "--platform=node" to do that, which will remove this error.

6 of 12 errors shown (disable the message limit with --log-limit=0)
 ELIFECYCLE  Command failed with exit code 1.
```

**Comment by philipp-spiess** [comment]
Nice! This is the same "solution" we came up with for the VS Code extension 🙃 https://github.com/sourcegraph/sourcegraph/pull/42802/files

I wonder if we should make this config extension wide (e.g. `cody.proxy` or so)? We just need to make sure we pass it to all fetch() calls but we don't have to do that immediatly

**Comment by philipp-spiess** [comment]
DMed you this as well but posting it here for completeness too.

The CI issues come beacuse `socks` is a node only library but VS Code extensions are generic to the underlying platform so they need to work on both node (desktop) and web (browser).

To fix this, we'll need to make sure to only include this library in the desktop build. We currently split based on these two entry files:

- https://github.com/sourcegraph/cody/blob/main/vscode/src/extension.node.ts
- https://github.com/sourcegraph/cody/blob/main/vscode/src/extension.web.ts

So you need to make sure that the code for this is injected at these endpoints and stubbed out for web.

File: agent/src/protocol.ts [hunk 1]
```diff
@@ -154,6 +154,7 @@ export interface ExtensionConfiguration {
     customHeaders: Record<string, string>
     autocompleteAdvancedProvider?: string
     autocompleteAdvancedServerEndpoint?: string | null
+    autocompleteAdvancedServerSocksProxy?: string | null
```
[the remark this answers is not in the document]
Comment (taras-yemets) [reply agent/src/protocol.ts:157]: Makes sense. In this case we use `autocompleteAdvancedServerSocksProxy` for `autocompleteAdvancedServerEndpoint`. `serverEndpoint` might have its own proxy. What do you think about having `proxies` as `Record<string, string` where we can map endpoints to proxies, e.g.
```
proxies: {
  [SERVER_ENDPOINT]: [SERVER_ENPOINT_PROXY],
  [AUTOCOMPLETE_ENDPOINT]: [AUTOCOMPLETE_ENDPOINT_PROXY]
}
```
Then in `fetch.node.ts` we can check whether there is a proxy for the requested URL and return the corresponding proxy agent. For example, if the proxy endpoint protocol is socks, we return the socks proxy agent instance.
cc: @chwarwick @philipp-spiess

File: vscode/package.json [hunk 2]
```diff
@@ -869,6 +869,10 @@
           "type": "string",
           "markdownDescription": "The server endpoint used for code autocomplete. This is only supported with a provider other than `anthropic`."
         },
+        "cody.autocomplete.advanced.serverSocksProxy": {
+          "type": "string",
+          "markdownDeprecationMessage": "The SOCKS proxy endpoint to access server endpoint. This is only supported with `unstable-codegen` provider and only for use with the Cody Agent (for instance with JetBrains plugin). For VS Code please use http.proxy instead."
```
[the remark this answers is not in the document]
Comment (taras-yemets) [reply vscode/package.json:874]: As we provide `markdownDeprecationMessage` this setting is not visible in the settings UI ([docs](https://code.visualstudio.com/api/references/contribution-points#contributes.configuration)). This setting is:
- not shown in settings UI
Screenshot 2023-09-07 at 10 33 11

- not suggested in autocomplete options in `settings.json`
Screenshot 2023-09-07 at 10 36 02

- can be typed manually with the deprecation message shown 
Screenshot 2023-09-07 at 10 32 01


Updated the description.

File: agent/src/protocol.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (taras-yemets) [reply agent/src/protocol.ts]: Renamed to "proxy" https://github.com/sourcegraph/cody/pull/836/commits/01f9c04566ea6890247a241367f8342404a684ff

[the remark this answers is not in the document]
Comment (taras-yemets) [reply agent/src/protocol.ts]: Corresponding change in the paired PR: https://github.com/sourcegraph/sourcegraph/pull/56264/commits/22f0dde3459ac914a94dbb89cc0286380cbea723

File: vscode/package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (chwarwick) [reply vscode/package.json]: Nice the picture in the description was what got me.

**Comment by slimsag** [comment]
Nice work on bringing this to the finish line @taras-yemets :D

File: vscode/src/fetch.node.ts [hunk 3]
```diff
@@ -1,29 +1,51 @@
...
+    return ({ protocol }) => {
+        if (proxy) {
+            if (proxy.startsWith('socks') && !socksProxyAgent) {
+                socksProxyAgent = new SocksProxyAgent(proxy, {
+                    keepAlive: true,
+                    keepAliveMsecs: 60000,
+                })
+                return socksProxyAgent
+            }
+        }
```
Comment (philipp-spiess) [inline vscode/src/fetch.node.ts:34]:
```suggestion
        if (proxy && proxy.startsWith('socks') && !socksProxyAgent) {
            socksProxyAgent = new SocksProxyAgent(proxy, {
                keepAlive: true,
                keepAliveMsecs: 60000,
            })
            return socksProxyAgent
        }
```

Is there a reason why the `proxy` branch with a misconfigured proxy won't result to the http agents with connection reuse?

**Review by philipp-spiess — APPROVED** [review]

File: vscode/src/fetch.node.ts (hunk 3 above)
Comment (taras-yemets) [reply vscode/src/fetch.node.ts:34]: Great catch! It was not intentional.
Thanks for your help and review, @philipp-spiess!

Comment (taras-yemets) [reply vscode/src/fetch.node.ts]: Addressed in https://github.com/sourcegraph/cody/pull/836/commits/dc38d86f177f445513b1a8353855f19fc391f582

**Pull request merged**